Add support for RX architecture CI on Zephyr

duynguyenxa opened the pull request:
This PR is a continued work on the https://github.com/zephyrproject-rtos/zephyr/pull/89770 to fix all RX architecture qemu environment to enable CI test for it.
These fixes will also need to apply change in the qemu source code, fix patch is updated at: https://github.com/zephyrproject-rtos/sdk-ng/pull/944

keith-packard committed:
arch/rx: Add missing linker script fragments

There were lots of missing chunks, including those necessary for C++
support.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests/arch/rx: Remove unused variable 'events'

This causes a compile error.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
subsys/debug: Stub out debug thread info for RX arch

This architecture is missing lots of support bits. Stub out this piece so
we can get more things building.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
scripts/logging: Allow targets that prefix global symbols with '_'

RX adds an underscore to all global symbols, so when we look for
log_const_ globals, we need to allow for that.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
scripts/logging: Skip log symbols of zero size

I'm not sure why this was necessary, but analyzing RX
executables was crashing with a divide by zero error on a
symbol which appeared to be duplicated in this list.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
scripts/logging: Add RX config detection

The RX arch uses the CONFIG_RX kconfig symbol

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
subsys/mgmt: Declare PROCESSOR_NAME for RX targets

Map the current set of RX CPU variants into descriptive names

Signed-off-by: Keith Packard <keithp@keithp.com>

duynguyenxa committed:
arch: rx: Add ARCH_EXCEPT for RX architect

Adding ARCH_EXCEPT for RX architect by using unconditional
trap interrupt no 3.

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

keith-packard committed:
compiler/gcc: Add rx compiler flags to linker too

Many of these affect the library search path.

Signed-off-by: Keith Packard <keithp@keithp.com>

duynguyenxa committed:
include: arch: rx: Change data symbol name

The RX linker is using the name "data" for .data region start
which caused the application cannot use "data" for variable
naming globally.
This commit change the name to "__data_region_start" as expected
in the linker_defs.h file

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
drivers: serial: Fix issue qemu rx cannot input console

The source code is missing the Receive enable in serial
poll-in function so the status flag will never become affect
and data will never be read
Add the RE write before checking for receive status flag

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
include: arch: Update missing rx/exception.h include

Add missing zephyr/arch/rx/exception.h for RX architecture

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
kernel: memslab:  Prevent pointer underflow in free list init

The "p -= slab->info.block_size;" is causing the "p" pointer
to be underflow in RX architecture case, where the RAM address
start from 0x0, in some case p minus block size make it underflow
This change implementation uses an index-based reverse loop to
safely iterate over each block from last to first.

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
tests: interrupt_util: Update RX irq trigger method

The QEMU RX doesn't allow to write value to IR flag,
There is also mentioned in RX HWM that the IR should
not be write 1 to. So we change the method to trigger
interrupt here is to call directly to SW irq table

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

Commit:
lib: posix: Update attribute packed for posix_thread_attr

Add attributed packed for struct posix_thread_attr when arch is RX

Signed-off-by: Minh Tang <minh.tang.ue@bp.renesas.com>

duynguyenxa committed:
include: arch: rx: Change noinit region to be after bss

Change noinit region to be after bss so that it would not cause
alignment issue for the data region setup

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
board: qemu_rx: Ignore net and bluetooth related test for qemu_rx

Add testing filter to ignore net and bluetooth test for qemu_rx

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
tests: libcxx: Update missing CONFIG_COMMON_LIBC_MALLOC

The prj.conf was missing CONFIG_COMMON_LIBC_MALLOC causing
warning of the depended config CONFIG_COMMON_LIBC_MALLOC_ARENA_SIZE

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
boards: qemu_rx: Add ignore can test

qemu_rx platform doesn't support can bus so we skip this test

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
tests: error_hook: Workaround for RX architecture

GCC for RX would crash with an internal compiler error during
DWARF frame generation in `dwarf2out_frame_debug_adjust_cfa()`
when compiling the `trigger_fault_illegal_instruction()`
test function.

This patch adds `__builtin_unreachable()` to help the compiler
reason about control flow, preventing incorrect DWARF CFA
generation.

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
arch: rx: Add simple sys_arch_reboot for RX architecture

Add simple sys_arch_reboot for RX architecture to conduct CI

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa commented:
Current result execute on local with qemu-target-rx fix

There are 6 failure cases seem to be issue with twister harness cannot send input command to qemu.
Others are still under checking

```

ERROR   - qemu_rx/r5f562n8          samples/sensor/sensor_shell/sample.sensor.shell.pytest FAILED: 5/5 pytest scenario(s) failed
ERROR   - see: /home/duyhoangnguyen/zephyrproject/zephyr/twister-out/qemu_rx_r5f562n8/zephyr/samples/sensor/sensor_shell/sample.sensor.shell.pytest/twister_harness.log
ERROR   - qemu_rx/r5f562n8          samples/basic/hash_map/sample.libraries.hash_map.newlib.cxx_unordered_map.djb2 FAILED: unexpected eof
ERROR   - see: /home/duyhoangnguyen/zephyrproject/zephyr/twister-out/qemu_rx_r5f562n8/zephyr/samples/basic/hash_map/sample.libraries.hash_map.newlib.cxx_unordered_map.djb2/build.log
ERROR   - qemu_rx/r5f562n8          samples/subsys/shell/shell_module/sample.shell.shell_module FAILED: 1/1 pytest scenario(s) failed
ERROR   - see: /home/duyhoangnguyen/zephyrproject/zephyr/twister-out/qemu_rx_r5f562n8/zephyr/samples/subsys/shell/shell_module/sample.shell.shell_module/twister_harness.log
ERROR   - qemu_rx/r5f562n8          samples/subsys/testsuite/pytest/shell/sample.harness.shell.vt100_colors_off FAILED: 1/1 pytest scenario(s) failed
ERROR   - see: /home/duyhoangnguyen/zephyrproject/zephyr/twister-out/qemu_rx_r5f562n8/zephyr/samples/subsys/testsuite/pytest/shell/sample.harness.shell.vt100_colors_off/twister_harness.log
ERROR   - qemu_rx/r5f562n8          samples/subsys/testsuite/pytest/shell/sample.harness.shell FAILED: 1/1 pytest scenario(s) failed
ERROR   - see: /home/duyhoangnguyen/zephyrproject/zephyr/twister-out/qemu_rx_r5f562n8/zephyr/samples/subsys/testsuite/pytest/shell/sample.harness.shell/twister_harness.log
ERROR   - qemu_rx/r5f562n8          samples/subsys/testsuite/pytest/shell/sample.pytest.shell FAILED: 2/2 pytest scenario(s) failed
ERROR   - see: /home/duyhoangnguyen/zephyrproject/zephyr/twister-out/qemu_rx_r5f562n8/zephyr/samples/subsys/testsuite/pytest/shell/sample.pytest.shell/twister_harness.log
ERROR   - qemu_rx/r5f562n8          samples/subsys/testsuite/pytest/shell/sample.pytest.shell.vt100_colors_off FAILED: 2/2 pytest scenario(s) failed
ERROR   - see: /home/duyhoangnguyen/zephyrproject/zephyr/twister-out/qemu_rx_r5f562n8/zephyr/samples/subsys/testsuite/pytest/shell/sample.pytest.shell.vt100_colors_off/twister_harness.log
ERROR   - qemu_rx/r5f562n8          tests/lib/hash_map/libraries.hash_map.cxx.djb2     FAILED: unexpected eof
ERROR   - see: /home/duyhoangnguyen/zephyrproject/zephyr/twister-out/qemu_rx_r5f562n8/zephyr/tests/lib/hash_map/libraries.hash_map.cxx.djb2/build.log
ERROR   - qemu_rx/r5f562n8          tests/lib/cpp/libcxx/cpp.libcxx.glibcxx.picolibc   FAILED: unexpected eof
ERROR   - see: /home/duyhoangnguyen/zephyrproject/zephyr/twister-out/qemu_rx_r5f562n8/zephyr/tests/lib/cpp/libcxx/cpp.libcxx.glibcxx.picolibc/build.log
ERROR   - qemu_rx/r5f562n8          tests/lib/cpp/libcxx/cpp.libcxx.glibcxx.newlib     FAILED: unexpected eof
ERROR   - see: /home/duyhoangnguyen/zephyrproject/zephyr/twister-out/qemu_rx_r5f562n8/zephyr/tests/lib/cpp/libcxx/cpp.libcxx.glibcxx.newlib/build.log
WARNING - A None status detected in instance qemu_rx/r5f562n8/zephyr/tests/subsys/fs/lib_link/filesystem.lib_link, test case filesystem.lib_link.
ERROR   - qemu_rx/r5f562n8          tests/subsys/zbus/integration/message_bus.zbus.module_interaction_no_error FAILED: Testsuite failed
ERROR   - see: /home/duyhoangnguyen/zephyrproject/zephyr/twister-out/qemu_rx_r5f562n8/zephyr/tests/subsys/zbus/integration/message_bus.zbus.module_interaction_no_error/handler.log
INFO    - Total complete:  777/ 777  100%  built (not run):   24, filtered: 2946, failed:   11, error:    0
INFO    - 3741 test scenarios (3519 configurations) selected, 2946 configurations filtered (2742 by static filter, 204 at runtime).
INFO    - 529 of 573 executed test configurations passed (92.32%), 24 built (not run), 11 failed, 0 errored, with 1 warnings in 721.10 seconds.
INFO    - 4527 of 4566 executed test cases passed (99.15%), 25 blocked, 13 failed, 1 without a status on 1 out of total 1088 platforms (0.09%).
INFO    - 624 selected test cases not executed: 588 skipped, 36 not run (built only).
INFO    - 540 test configurations executed on platforms, 15 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/duyhoangnguyen/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/duyhoangnguyen/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/duyhoangnguyen/zephyrproject/zephyr/twister-out/twister_report.xml...
```

nashif commented on boards/qemu/rx/qemu_rx.yaml:
```diff
@@ -9,7 +9,15 @@ simulation:
 toolchain:
   - cross-compile
+  - zephyr
+  - zephyr-gnu
```
zephyr-gnu is not needed and is not used anymore in the sdk, just drop it for now.

duynguyenxa committed:
dts: rx: Fix incorrect ram size information for RX

The RX QEMU platform SRAM size is 96KB, this commit correct the
device tree information

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa commented:
Update: Incorrect define of SRAM size in dts of qemu_rx make tests/lib/cpp/libcxx/cpp.libcxx.glibcxx.newlib due to not enough RAM, I updated to correct size, now test case is passed.
Remaining 6 failed due to twister harness and 1 unknown reason failed case: tests/subsys/zbus/integration/message_bus.zbus.module_interaction_no_error

duynguyenxa committed:
test: zbus: Disable icount shift on qemu_rx

Somehow icount shift making this test fail on qemu_rx target
maybe clock rate on RX is too low, as work around I disable
the icount shift for this test case on qemu_rx

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
samples: Add exclusion for twister harness on qemu_rx

The qemu_rx is having issue with twister harness, command cannot
be send from twister test to qemu console, this commit temporary
disable them for further checking

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
test/subsys/jwt: Work around for qemu_rx support

The qemu_rx some how with too big offset in "mov" instruction
cannot manipulate the memory correctly
This commit reconfigure the slot count to reduce the size of
psa_global_data_t struct so that issue on qemu_rx not occur

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
boards: qemu: Add zephyr toolchain to qemu_rx

Add Zephyr toolchain to qemu_rx boards to support CI

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
boards: renesas: Add zephyr toolchain to rsk_rx130 board

Add zephyr toolchain support to rsk_rx130 board

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
dts: rx: Update interrupt-controller node

Update interrupt controller node to be compatible with new
binding configuration

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

nashif commented:
@duynguyenxa  can you please rebase, we now have qemu 10.0.2 with RX patches in the SDK/docker. Lets see how things work out.

kartben commented:
> @duynguyenxa can you please rebase, we now have qemu 10.0.2 with RX patches in the SDK/docker. Lets see how things work out.

re-running CI should be enough - I just closed/open the PR to do just that

stephanosio commented:
> re-running CI should be enough - I just closed/open the PR to do just that

It needs a rebase. Twister workflow is no longer `pull_request_target`.

duynguyenxa committed:
github: hello_world_multiplatform: Adding rx-zephyr-elf

Adding rx-zephyr-elf for testing rx target

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

nashif commented:
> @nashif , yes it's due to #89555 merge change the binding of ICU node of RX fix is provide at [f2fca08](https://github.com/zephyrproject-rtos/zephyr/commit/f2fca08cb8ead0b822f867bf4cc617fbf3c6bc58) in this PR. Sorry, should've finished this first before the other PR merge

ok, but still when it builds now, it does not run at all in qemu.

duynguyenxa commented:
@nashif , sorry this is my mistake, I'm adding the 00019 patch but forget to put it in the bb the sdk, filehttps://github.com/zephyrproject-rtos/sdk-ng/blob/v0.17-branch/meta-zephyr-sdk/recipes-devtools/qemu/qemu-zephyr_10.0.2.bb

nashif commented:
> @nashif , sorry this is my mistake, I'm adding the 00019 patch but forget to put it in the .bb file, the qemu in Zephyr SDK is missing 1 patch
> [zephyrproject-rtos/sdk-ng@`v0.17-branch`/meta-zephyr-sdk/recipes-devtools/qemu/qemu-zephyr_10.0.2.bb](https://github.com/zephyrproject-rtos/sdk-ng/blob/v0.17-branch/meta-zephyr-sdk/recipes-devtools/qemu/qemu-zephyr_10.0.2.bb)

ok, partly also my mistake, should have caught that. Any way we can work around this, the offending commit right now is 8ddb13d82ad, can we try and get at least the basic testing with qemu_rx working without the need for another SDK release?

nashif commented:
> yes we can, Only need to drop [8ddb13d](https://github.com/zephyrproject-rtos/zephyr/commit/8ddb13d82adeb04994d5de6aa981e701fc017490) and [c85e12b](https://github.com/zephyrproject-rtos/zephyr/commit/c85e12bbf45544a38ae12da0d6874593651da85a) in this PR, but we have to add exclude for some cases because of bad bin generate by objcopy.

Ok, maybe remove those 2 commits for now and do the following in the board yaml file:

```
testing:
  default: true
  only_tags:
     - kernel
  ignore_tags:
    - net
    - bluetooth
    - can
 ```

and get this going, you can later remove this and fine tune filtering.

duynguyenxa committed:
CMakeList: Disable gapfill for RX target

Ignore gapfill for RX target as some RX toolchains generate wrong
output image when running gapfill for binary format.

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa commented:
> > yes we can, Only need to drop [8ddb13d](https://github.com/zephyrproject-rtos/zephyr/commit/8ddb13d82adeb04994d5de6aa981e701fc017490) and [c85e12b](https://github.com/zephyrproject-rtos/zephyr/commit/c85e12bbf45544a38ae12da0d6874593651da85a) in this PR, but we have to add exclude for some cases because of bad bin generate by objcopy.
> 
> Ok, maybe remove those 2 commits for now and do the following in the board yaml file:
> 
> ```
> testing:
>   default: true
>   only_tags:
>      - kernel
>   ignore_tags:
>     - net
>     - bluetooth
>     - can
> ```
> 
> and get this going, you can later remove this and fine tune filtering.

I drop the 2 commits, and disable the gap_fill option on RX target for now

duynguyenxa committed:
boards: renesas: rsk_rx130: Add missing ioport node define

Board device tree is missing ioport3 status set okay for
button. Add missing ioport3 enable

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa commented:
> one failing test is due to #92187

I checked the other on my local but it's Passed without changing anything, now not sure how I can solve this.

![image](https://github.com/user-attachments/assets/ca1ca554-a7e1-4b68-9c34-2fc46924f326)

duynguyenxa committed:
tests: kernel_workq: Excluding test for qemu_rx

Excluding this test for qemu_rx because it's failing randomly
Root cause not identified yet, #92213 is created to keep track
of this
This test case is working fine on actual RX HW so issue might be
in the inside the qemu environment

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
boards: renesas: Doc update for RSK-RX130 boards with SDK 0.17.1

From Zephyr SDK 0.17.1 the RX architecture is supported so
update the documentation of RX130 board to reflect this

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa commented:
It seem that after a few time re-attempt the job pass, it's passing inconsistently.
I verified the test case on actual HW the RSK-RX130 many time and there are no issue, maybe there is issue in qemu environment, anyway I created the https://github.com/zephyrproject-rtos/zephyr/issues/92213 to track it, hopefully can resolve it soon
In this PR I exclude the test case kernel.workqueue.api for qemu RX
